migrator/main.go

```diff
@@ -56,38 +61,100 @@ func run() error {
+	// Scan the RocksDB replicas
+	err := dbmRocks.Scan()
+	if err != nil {
+		return errors.Wrap(err, "fail to scan replicas")
+	}
 
-		replicaName, replicaPath, err := dbm.GetReplicaToMigrate()
+	replica, replicaPath, err := dbmRocks.GetReplicaToMigrate()
+
+	// We are assuming that Rocks is record of truth if current replica is present.  So in these cases
+	// we need to process RocksDB migrations first.
+	if !features.PostgresDatastore.Enabled() || (features.PostgresDatastore.Enabled() && replica == rocksdb.CurrentReplica) {
+		log.WriteToStderrf("Rocks replica to process = %q", replica)
 		if err != nil {
 			return err
 		}
 		option.MigratorOptions.DBPathBase = replicaPath
-		if err = upgrade(conf); err != nil {
+
+		err = upgrade(conf, "", true)
+		if err != nil {
 			return err
 		}
-		if err = dbm.Persist(replicaName); err != nil {
+
+		if err = dbmRocks.Persist(replica); err != nil {
 			return err
 		}
+		makeEmergencyReplica = true
 	} else {
-		var gormDB *gorm.DB
-		gormDB, err := postgreshelper.Load(conf)
+		log.WriteToStderrf("Not persisting Rocks replica = %q", replica)
+	}
+
+	if features.PostgresDatastore.Enabled() {
+		sourceMap, adminConfig, err := pgconfig.GetPostgresConfig()
+		if err != nil {
+			return errors.Wrap(err, "unable to get Postgres DB config.")
+		}
+		dbm := postgres.New(conf.Maintenance.ForceRollbackVersion, adminConfig, sourceMap)
+
+		// Scan for database replicas
+		err = dbm.Scan()
+		if err != nil {
+			return errors.Wrap(err, "fail to scan replicas")
+		}
+
+		replica, _, err := dbm.GetReplicaToMigrate()
+		log.WriteToStderrf("Replica to migrate => %q", replica)
+		if err != nil {
+			return err
+		}
+
+		// No existing DB replica so need to create it.
+		gormDB, err := postgreshelper.Load(conf, replica)
 		if err != nil {
 			return errors.Wrap(err, "failed to connect to postgres DB")
 		}
 		pkgSchema.ApplyAllSchemas(context.Background(), gormDB)
-		log.WriteToStderr("Applied all table schemas.")
+
+		// We need to close the gorm connection when we are finished with it so the replica
+		// movement can occur.
+		sqlDB, err := gormDB.DB()
+		if err != nil {
+			log.WriteToStderrf("Error closing GormDB connection: %v", err)
+		}
+		defer func() {
+			if err := sqlDB.Close(); err != nil {
+				log.WriteToStderrf("Error closing DB: %v", err)
+			}
+		}()
+
+		// Run upgrades
+		if err = upgrade(conf, replica, false); err != nil {
+			return err
+		}
+
+		// Save the replica after upgrades have been processed
+		if err = dbm.Persist(replica); err != nil {
+			return err
+		}
+
+		// We have upgraded from RocksDB to Postgres.  We need to make an emergency fallback copy of
+		// RocksDB in the event we need to rollback.
+		if makeEmergencyReplica {
+			if err = dbmRocks.MakeEmergencyReplica(); err != nil {
+				return err
+			}
+		}
 	}
 
 	return nil
 }
 
-func upgrade(conf *config.Config) error {
+func upgrade(conf *config.Config, dbReplica string, runLegacy bool) error {
```

Comment:
The idea there is a case where we are doing an upgrade from sometime back and upgrading to Postgres at the same time.  Thought was first run through the migrations on rocks to get it up to date and then migrate the updated rocks to postgres.  So no, I can't just use the flag as I need to update Rocks first.

Reply:
We do not have to seperate the process into two phrases. The DB sequence number should tell us the sequence number to start and end.

Reply:
I was going to change it so update is only called once and the runner figures out what to run.  If the runner recives both a rocks and a postgres it will have to determine what to run and in which order.  That just can't fully happen until we wire in your migrations.

Reply:
We are going to have to push some of this to the various tickets dealing with wiring in the migrations.  I can't handle all that with the replicas.  I can send the right replicas with the right versions to the migrations runner, but I think that is as far as I can go in this PR.